src/builtins.c

```diff
@@ -411,13 +415,21 @@ static uintptr_t NOINLINE jl_object_id__cold(jl_datatype_t *dt, jl_value_t *v) J
     }
     if (dt->name->mutabl)
         return inthash((uintptr_t)v);
-    return immut_id_(dt, v, dt->hash);
+    return immut_id_(dt, v, dt->hash, object_table);
 }
 
-JL_DLLEXPORT inline uintptr_t jl_object_id_(jl_value_t *tv, jl_value_t *v) JL_NOTSAFEPOINT
+JL_DLLEXPORT inline uintptr_t jl_object_id_(jl_value_t *tv, jl_value_t *v, htable_t *object_table) JL_NOTSAFEPOINT
 {
+    if(object_table == NULL) 
+        return jl_object_id(v);
+    if(v != NULL) {
+        uintptr_t *hashp = (uintptr_t*)ptrhash_bp(object_table, v);
```

Comment:
@vtjnash, I added an extra check to ensure that `v` is not `NULL` before getting it's value from the `object_table`, so that memory that doesn't exist is not accessed. I did that while inserting it into the table as well. However, I'm getting a SIGABRT from `toml_parser.jl`. It is apparently because of the following lines of code:
```julia
const err_message = Dict(
    ErrTrailingCommaInlineTable             => "trailing comma not allowed in inline table",
    ErrExpectedCommaBetweenItemsArray       => "expected comma between items in array",
    ErrExpectedCommaBetweenItemsInlineTable => "expected comma between items in inline table",
    ErrExpectedEndArrayOfTable              => "expected array of table to end with ']]'",
    ErrInvalidBareKeyCharacter              => "invalid bare key character",
    ErrRedefineTableArray                   => "tried to redefine an existing table as an array",
    ErrDuplicatedKey                        => "key already defined",
    ErrKeyAlreadyHasValue                   => "key already has a value",
    ErrEmptyBareKey                         => "bare key cannot be empty",
    ErrExpectedNewLineKeyValue              => "expected newline after key value pair",
    ErrNewLineInString                      => "newline character in single quoted string",
    ErrUnexpectedEndString                  => "string literal ened unexpectedly",
    ErrExpectedEndOfTable                   => "expected end of table ']'",
    ErrAddKeyToInlineTable                  => "tried to add a new key to an inline table",
    ErrArrayTreatedAsDictionary             => "tried to add a key to an array",
    ErrAddArrayToStaticArray                => "tried to append to a statically defined array",
    ErrGenericValueError                    => "failed to parse value",
    ErrLeadingZeroNotAllowedInteger         => "leading zero in integer not allowed",
    ErrUnderscoreNotSurroundedByDigits      => "underscore is not surrounded by digits",
    ErrUnexpectedStartOfValue               => "unexpected start of value",
    ErrOffsetDateNotSupported               => "offset date-time is not supported",
    ErrParsingDateTime                      => "parsing date/time value failed",
    ErrTrailingUnderscoreNumber             => "trailing underscore in number",
    ErrLeadingDot                           => "floats require a leading zero",
    ErrExpectedEqualAfterKey                => "expected equal sign after key",
    ErrNoTrailingDigitAfterDot              => "expected digit after dot",
    ErrOverflowError                        => "overflowed when parsing integer",
    ErrInvalidUnicodeScalar                 => "invalid unicode scalar",
    ErrInvalidEscapeCharacter               => "invalid escape character",
    ErrUnexpectedEofExpectedValue           => "unexpected end of file, expected a value"
)
```

Even the `make debug` command gets stuck at `toml_parser.jl`.

Reply:
Ah, this is subtle, but 'v' is not a unique identifier, since it could be either the object or the first field. We need to be careful to specifically only use the cache if we know it is the

Reply:
@vtjnash, I'm sorry, but I don't understand what you mean by `v` can be either an object or a field. What exactly do you mean by a field?

Reply:
If the object looks like this: 
```
struct jl_value_t {
    int field1;
} *v;
```
So `v == &v->field1` but `hash(v) != hash(v->field1)`

Reply:
So this function can only use the ptrhash when it has `v` when it is known to be directly from `jl_object_id_temp`, and not when this function gets used directly

Reply:
So are you suggesting that I should retrieve the object from the cache in `jl_object_id_temp` instead of `jl_object_id_` function? Are you suggesting somehow the value of `v` changes to `v->field` when passing it to the `jl_object_id_` function from the `jl_object_id_temp` function? And why should the same not happen when calling `jl_object_id_temp` function?

Reply:
I could also check if the `object_table` is null right? Because the only way it isn't null is if it's called from `jl_object_id` function. Also, `jl_object_id` is called in a couple more places, such as in `codegen.cpp`. How do I change its function call?

Reply:
Ah, okay, I see now I was misleading you. There are not other calls to it here, just that one in codegen that you noted. Here is how to add an argument to that call:
```diff
diff --git a/src/ccall.cpp b/src/ccall.cpp
index ce240b86e7..dcf8bc7bcc 100644
--- a/src/ccall.cpp
+++ b/src/ccall.cpp
@@ -1801,7 +1801,8 @@ static jl_cgval_t emit_ccall(jl_codectx_t &ctx, jl_value_t **args, size_t nargs)
                 val.isghost ? ConstantPointerNull::get(T_pint8_derived) :
                     ctx.builder.CreateBitCast(
                         decay_derived(ctx, data_pointer(ctx, val)),
-                        T_pint8_derived)
+                        T_pint8_derived),
+                ConstantPointerNull::get(getInt8PtrTy(ctx.builder.getContext()))
             };
             Value *ret = ctx.builder.CreateCall(prepare_call(jl_object_id__func), makeArrayRef(args));
             JL_GC_POP();
diff --git a/src/codegen.cpp b/src/codegen.cpp
index b7483d1e88..598450aa77 100644
--- a/src/codegen.cpp
+++ b/src/codegen.cpp
@@ -897,7 +897,7 @@ static const auto jlapplytype_func = new JuliaFunction{
 static const auto jl_object_id__func = new JuliaFunction{
     XSTR(jl_object_id_),
     [](LLVMContext &C) { return FunctionType::get(getSizeTy(C),
-            {JuliaType::get_prjlvalue_ty(C), PointerType::get(getInt8Ty(C), AddressSpace::Derived)}, false); },
+            {JuliaType::get_prjlvalue_ty(C), PointerType::get(getInt8Ty(C), AddressSpace::Derived), getInt8PtrTy(C)}, false); },
     nullptr,
 };
 static const auto setjmp_func = new JuliaFunction{
```

Reply:
Because I had changed the definition of `jl_object_id_` function in `julia_internal.h` to include the hash_table as an argument. So any function call to `jl_object_id_` should pass a pointer to the hash function. My idea was that if I pass `NULL` as the `object_table` argument to the `jl_object_id_` function, then whenever the `jl_object_id_` function is called directly, it will not use the cache. And when it is called from the `jl_object_id_temp` function, then the `object_table` won't be `NULL` and so we can use the cache then.

Reply:
This call is not type-checked against the function, so it would pass a random value there